README.md
# Carrierwave::Qiniu

[![Gem Version](https://badge.fury.io/rb/carrierwave-qiniu@2x.png?1.2.2)](http://badge.fury.io/rb/carrierwave-qiniu)

This gem adds storage support for [Qiniu](http://qiniutek.com) to [Carrierwave](https://github.com/jnicklas/carrierwave)

example: https://github.com/huobazi/carrierwave-qiniu-example

## Installation

Add the following to your application's Gemfile:

    gem 'carrierwave-qiniu', '~> 1.2.2'
    gem 'carrierwave-i18n' # If you need to use locales other than English

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install carrierwave-qiniu -v 1.2.2

## Usage

You'll need to configure it in config/initializers/carrierwave.rb

```ruby
::CarrierWave.configure do |config|
  config.storage              = :qiniu
  config.qiniu_access_key     = "your qiniu access_key"
  config.qiniu_secret_key     = 'your qiniu secret_key'
  config.qiniu_bucket         = "carrierwave-qiniu-example"
  config.qiniu_bucket_domain  = "carrierwave-qiniu-example.aspxboy.com"
  config.qiniu_bucket_private = true #default is false
  config.qiniu_block_size     = 4*1024*1024
  config.qiniu_protocol       = "http"
  config.qiniu_up_host        = 'http://up.qiniug.com' #七牛上传海外服务器,国内使用可以不要这行配置
end
```

For more information on qiniu, please read http://developer.qiniu.com/docs/v6/

And then in your uploader, set the storage to `:qiniu`:

```ruby
class AvatarUploader < CarrierWave::Uploader::Base
  storage :qiniu
end
```

You can override configuration item in individual uploader like this:

```ruby
class AvatarUploader < CarrierWave::Uploader::Base
  storage :qiniu

  self.qiniu_bucket                = "avatars"
  self.qiniu_bucket_domain         = "avatars.files.example.com"
  self.qiniu_protocal              = 'http'
  self.qiniu_delete_after_days     = 30
  self.qiniu_can_overwrite         = true
  self.qiniu_bucket_private        = true #default is false
  self.qiniu_callback_url          = "http://<ip1>/callback;http://<ip2>/callback"
  self.qiniu_callback_body         = "key=$(key)&hash=$(etag)&w=$(imageInfo.width)&h=$(imageInfo.height)" # see http://developer.qiniu.com/docs/v6/api/overview/up/response/vars.html#magicvar
  self.qiniu_persistent_notify_url = "http://<ip>/notify"
  
  # 使用的队列名称,不设置代表不使用私有队列，使用公有队列
  # 可以自己创建私有队列 see: https://portal.qiniu.com/dora/mps/new
  self.qiniu_persistent_pipeline = "marble_persistent"
    
  # 指定预转数据处理命令,返回由每条命令组成的数组
  # See
  # https://developer.qiniu.com/dora/api/1291/persistent-data-processing-pfop#
  # https://developer.qiniu.com/dora/api/3686/pfop-directions-for-use
  # https://developer.qiniu.com/kodo/manual/1206/put-policy#persistentOps
  def qiniu_persistent_ops
    commands = []
    
    # 以预转持久化形式，将mp4视频转换为flv格式。
    # https://developer.qiniu.com/dora/api/1248/audio-and-video-transcoding-avthumb
    fops1 = "avthumb/flv"
    saveas_key1 = ::Qiniu::Utils.urlsafe_base64_encode("#{self.qiniu_bucket}:#{store_dir}/#{self.filename}_flv.flv")
    fops1 = fops1 + '|saveas/' + saveas_key1
    commands << fops1
    
    # 以预转持久化形式，将mp4视频转换为avi格式。
    # https://developer.qiniu.com/dora/api/1248/audio-and-video-transcoding-avthumb
    fops2 = "avthumb/avi"
    saveas_key2 = ::Qiniu::Utils.urlsafe_base64_encode("#{self.qiniu_bucket}:#{store_dir}/#{self.filename}_avi.avi")
    fops2 = fops2 + '|saveas/' + saveas_key2
    commands << fops2
    
    # 要进行视频截图操作。
    # https://developer.qiniu.com/dora/api/1313/video-frame-thumbnails-vframe
    fops3 = "vframe/jpg/offset/3/w/1280/h/720/rotate/auto"
    saveas_key3 = ::Qiniu::Utils.urlsafe_base64_encode("#{self.qiniu_bucket}:#{store_dir}/#{self.filename}_screenshot.jpg")
    fops3 = fops3 + '|saveas/' + saveas_key3
    commands << fops3
    
    commands
  end
end
```

You can use [qiniu image styles](https://qiniu.kf5.com/hc/kb/article/68884/) instead [version](https://github.com/carrierwaveuploader/carrierwave#adding-versions) processing of CarrierWave.

```ruby
# Case 1: Array styles
CarrierWave.configure do |config|
  config.qiniu_styles = [:thumb, :large]
end

class AvatarUploader < CarrierWave::Uploader::Base
  storage :qiniu

  use_qiniu_styles
end

# original url
user.avatar.url

# thumb url
user.avatar.url(:thumb)
# http://.../avatar.jpg-thumb


# Case 2: Hash styles
CarrierWave.configure do |config|
  config.qiniu_styles = { thumb: 'imageView2/1/w/200', large: 'imageView2/1/w/800' }
end

# thumb url
user.avatar.url(:thumb)
# http://.../avatar.jpg-thumb

# inline thubm url
user.avatar.url(:thumb, inline: true)
# http://.../avatar.jpg?imageView2/1/w/200

# just style param
user.avatar.url(style: 'imageView2/1/w/200')
# http://.../avatar.jpg?imageView2/1/w/200

# Case 3: Inline all styles in development environment
CarrierWave.configure do |config|
  config.qiniu_styles = { thumb: 'imageView2/1/w/200', large: 'imageView2/1/w/800' }
  config.qiniu_style_inline = true if Rails.env.development?
end

class AvatarUploader < CarrierWave::Uploader::Base
  storage :qiniu
  use_qiniu_styles
end

user.avatar.url(:thumb)
# http://.../avatar.jpg?imageView2/1/w/200

# Case 4: Custom styles and bucket
class AvatarUploader < CarrierWave::Uploader::Base
  storage :qiniu

  # Override default styles and use your own
  use_qiniu_styles :thumb => 'imageView/0/w/400', :xlarge => 'imageView/0/w/1600'

  self.qiniu_bucket        = "avatars"
  self.qiniu_bucket_domain = "avatars.files.example.com"

end

user.avatar.url(:thumb, inline: true)
# http://.../avatar.jpg?imageView2/1/w/400

```
Sync Qiniu styles of uploader

```
$ rake carrierwave:qiniu:sync_styles

# Bucket: bucket_name_1, Set style: thumb => imageView2/1/w/200
# Bucket: bucket_name_2, Set style: large => imageView2/1/w/800

```
You can see a example project on: https://github.com/huobazi/carrierwave-qiniu-example

or see the spec test on https://github.com/huobazi/carrierwave-qiniu/blob/master/spec/carrierwave-qiniu_spec.rb

## LICENSE

See the [LICENSE](https://github.com/huobazi/carrierwave-qiniu/blob/master/LICENSE.txt).

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Contributors

See the [Contributors List](https://github.com/huobazi/carrierwave-qiniu/graphs/contributors).

## CHANGE LOG

See the [CHANGELOGS.md](https://github.com/huobazi/carrierwave-qiniu/blob/master/CHANGELOG.md).


CHANGELOG.md

## CHANGE LOG

### v1.2.2

- 重构

### v1.2.1

- Implement of cache* method for carrierwave 2.0

### v1.2.0

- Fit to carrierwave 2.0 (支持 CarrierWave 2.0)

### v1.1.8

- Fix warning in ruby 2.7.

### v1.1.7

- Upgrade to rails 6 and qiniu 6.9

https://github.com/huobazi/carrierwave-qiniu/pull/85

### v1.1.6

- Add filename to storage qiniu file

https://github.com/huobazi/carrierwave-qiniu/pull/84

### v1.1.5

- 解决未定义的 qiniu_delete_after_days 和 qiniu_persistent_pipeline 方法错误

https://github.com/huobazi/carrierwave-qiniu/pull/81/

### v1.1.4

- 支持带样式的私有链接

https://github.com/huobazi/carrierwave-qiniu/pull/80

### v1.1.3

新增参数 qiniu_delete_after_days

参考: https://developer.qiniu.com/kodo/api/1732/update-file-lifecycle

### v1.1.2

Fix:
https://github.com/huobazi/carrierwave-qiniu/pull/79

### v1.1.1

https://github.com/huobazi/carrierwave-qiniu/pull/78

### v1.1.0

- Require carrierwave ~> 1.0
 
https://github.com/huobazi/carrierwave-qiniu/pull/73

### v1.0.1

- 使用 module ClassMethods 以兼容 Rails4

https://github.com/huobazi/carrierwave-qiniu/issues/71

https://github.com/huobazi/carrierwave-qiniu/pull/72

### v1.0.0

- 变更图片样式的用法
https://github.com/huobazi/carrierwave-qiniu/pull/70

- 不兼容上一版本

``` qiniu_styles ``` 移动到 config 中
在 uploader 内 使用 ``` use_qiniu_styles ``` 来指定使用默认 styles 或者 覆盖默认配置
详见 ReadMe 中的示例

### v0.2.6

- 提供图片样式的便利方法
https://github.com/huobazi/carrierwave-qiniu/pull/68


### v0.2.5

https://github.com/huobazi/carrierwave-qiniu/pull/61

https://github.com/huobazi/carrierwave-qiniu/pull/62


### v0.2.4

- Add test for upload failed
https://github.com/huobazi/carrierwave-qiniu/pull/57

### v0.2.3

- Add extension method
https://github.com/huobazi/carrierwave-qiniu/pull/50

### v0.2.2

- 增加对七牛SDK中复制图片功能的支持
https://github.com/huobazi/carrierwave-qiniu/pull/49

### v0.2.1

- 增加
```
    qiniu_callback_url
    qiniu_callback_body
    qiniu_can_overwrite
```

https://github.com/huobazi/carrierwave-qiniu/pull/47

### v0.2.0

- 升级七牛 SDK 至 6.5.1

### v0.1.8.2

https://github.com/huobazi/carrierwave-qiniu/pull/42

### v0.1.8.1

https://github.com/huobazi/carrierwave-qiniu/pull/36

https://github.com/huobazi/carrierwave-qiniu/pull/38

### v0.1.7

- 添加从云端读取（下载）文件  。 [https://github.com/huobazi/carrierwave-qiniu/pull/27](https://github.com/huobazi/carrierwave-qiniu/pull/27)

- 增加获取有时效性url链接地址的方法 。 [https://github.com/huobazi/carrierwave-qiniu/issues/25](https://github.com/huobazi/carrierwave-qiniu/issues/25)

- 升级七牛 SDK 至 6.4.2


lib/carrierwave/qiniu/style.rb
require 'carrierwave/qiniu/url'

module CarrierWave
  module Qiniu
    module Style
      extend ActiveSupport::Concern

      module ClassMethods
        # === Examples:
        #
        #    CarrierWave.configure do |config|
        #      config.qiniu_styles = [:thumbnail, :large]
        #      # or
        #      config.qiniu_styles = {:thumbnail => 'imageView/0/w/200', :large => 'imageView/0/w/800'}
        #    end
        #
        #    # Eanble qiniu styles otherwise default version processing
        #    # And use default styles
        #    use_qiniu_styles
        #
        #    # Override default styles and use your own styles
        #    use_qniu_styles :thumbnail => 'imageView/0/w/400', :xlarge => 'imageView/0/w/1600'
        #
        def use_qiniu_styles(versions = nil)

          # Override #url method when set styles, otherwise still default strategy.
          unless include? ::CarrierWave::Qiniu::Url
            send(:include, ::CarrierWave::Qiniu::Url)
          end

          @_qiniu_styles = {}
          if self.qiniu_styles
            # Set default styles
            @_qiniu_styles = parse_qiniu_styles(self.qiniu_styles)
          elsif versions
            # Set custom styles
            self.qiniu_styles = versions
            @_qiniu_styles = parse_qiniu_styles(versions)
          end
        end

        def get_qiniu_styles
          @_qiniu_styles
        end

        private
        def parse_qiniu_styles(styles)
          if styles.is_a? Array
            styles.map { |version| [version.to_sym, nil] }.to_h
          elsif styles.is_a? Hash
            styles.symbolize_keys
          end
        end
      end
    end
  end
end
